tests/transport/spdp/spdp_transport.cpp

```diff
@@ -0,0 +1,501 @@
+// Test for the mini-transport in SPDP (sequence number handling)
+
+#include "dds/DCPS/transport/rtps_udp/RtpsUdpInst.h"
+#ifdef ACE_AS_STATIC_LIBS
+#include "dds/DCPS/transport/rtps_udp/RtpsUdp.h"
+#endif
+
+#include "dds/DCPS/transport/framework/NetworkAddress.h"
+
+#include "dds/DCPS/RTPS/BaseMessageTypes.h"
+#include "dds/DCPS/RTPS/GuidGenerator.h"
+#include "dds/DCPS/RTPS/MessageTypes.h"
+#include "dds/DCPS/RTPS/RtpsCoreTypeSupportImpl.h"
+#include "dds/DCPS/RTPS/RtpsDiscovery.h"
+#include "dds/DCPS/RTPS/ParameterListConverter.h"
+#include "dds/DCPS/RTPS/Spdp.h"
+
+#include "dds/DCPS/Service_Participant.h"
+
+#include "ace/Configuration.h"
+#include "ace/Reactor.h"
+#include "ace/Select_Reactor.h"
+
+#include <ace/OS_main.h>
+#include <ace/Thread_Manager.h>
+#include <ace/Reactor.h>
+#include <ace/SOCK_Dgram.h>
+
+#include <exception>
+#include <iostream>
+
+using namespace OpenDDS::DCPS;
+using namespace OpenDDS::RTPS;
+
+// Declared as spdp_friend. Used to Interact Directly with Spdp.
+class DDS_TEST {
+public:
+  DDS_TEST(const RcHandle<Spdp>& spdp, const GUID_t& guid)
+  : spdp_(spdp)
+  , guid_(make_guid(guid.guidPrefix, ENTITYID_PARTICIPANT))
+  {
+  }
+
+  bool check_for_participant(bool expect_participant)
+  {
+    ACE_GUARD_RETURN(ACE_Thread_Mutex, g, spdp_->lock_, true);
+    const bool error = spdp_->has_discovered_participant(guid_) != expect_participant;
+    if (error) {
+      ACE_ERROR((LM_ERROR, ACE_TEXT("Error: %C\n"), expect_participant ?
+        "Expected to find the participant, but didn't" :
+        "Expected not to find the participant, but did"));
+    }
+    return error;
+  }
+
+  void remove_participant()
+  {
+    ACE_GUARD(ACE_Thread_Mutex, g, spdp_->lock_);
+    Spdp::DiscoveredParticipantIter iter = spdp_->participants_.find(guid_);
+    if (iter != spdp_->participants_.end()) {
+      spdp_->remove_discovered_participant(iter);
+    }
+  }
+
+  bool get_multicast_address(ACE_INET_Addr& addr) {
+    if (!spdp_->tport_) {
+      ACE_ERROR((LM_ERROR, ACE_TEXT("ERROR: Null SPDP Transport\n")));
+      return false;
+    }
+    OPENDDS_SET(ACE_INET_Addr)& send_addrs = spdp_->tport_->send_addrs_;
```

Comment:
I'm not sure about this assumption, especially with concurrent work ongoing with RtpsRelay.  Instead of checking `send_addrs_`, I suggest having the SpdpTransport constructor store the multicast address directly as a data member.